Remove deprecated code in the router

ocombe opened the pull request:
## PR Type
What kind of change does this PR introduce?

```
[ ] Other... Please describe: removing deprecated code
```

## What is the current behavior?
The values `true`, `false`, `legacy_enabled` and `legacy_disabled` for the router parameter `initialNavigation` are deprecated.
`RouterOutlet` properties `locationInjector` and `locationFactoryResolver` are deprecated since v4.

## What is the new behavior?
the values `true`, `false` for the router parameter `initialNavigation` have been removed as they were deprecated. Use `enabled` or `disabled` instead. The values `legacy_enabled` and `legacy_disabled` are still available but they will log a warning in dev mode and will be removed in v6.
`RouterOutlet` properties `locationInjector` and `locationFactoryResolver` have been removed as they were deprecated since v4.

## Does this PR introduce a breaking change?
```
[x] Yes
```

ocombe committed:
refactor(router): remove deprecated `initialNavigation` options

BREAKING CHANGE: the values `true`, `false` for the router parameter `initialNavigation` have been removed as they were deprecated. Use `enabled` or `disabled` instead. The values `legacy_enabled` and `legacy_disabled` are still available but they will log a warning in dev mode and will be removed in v6.

mary-poppins commented:
You can preview ff83f1b at https://pr18781-ff83f1b.ngbuilds.io/.

vicb commented on packages/router/src/router_module.ts:
```diff
@@ -332,9 +322,8 @@ export class RouterInitializer {
 
-      if (this.isLegacyDisabled(opts) || this.isLegacyEnabled(opts)) {
+      if (opts.initialNavigation === undefined) {
         resolve(true);
```
Looks wrong to me - should be the same as `enabled`.
Also the API docs need to be update to state that enabled is the default.

vicb commented on packages/router/src/router_module.ts:
```diff
@@ -372,26 +361,15 @@ export class RouterInitializer {
     }
 
-    if (this.isLegacyEnabled(opts)) {
+    if (opts.initialNavigation === undefined) {
```
same comment

mary-poppins commented:
You can preview 189d7d7 at https://pr18781-189d7d7.ngbuilds.io/.

jasonaden reviewed (approved):
LGTM

mary-poppins commented:
You can preview f85b04f at https://pr18781-f85b04f.ngbuilds.io/.

mary-poppins commented:
You can preview 1026836 at https://pr18781-1026836.ngbuilds.io/.

ocombe committed:
refactor(router): remove deprecated `RouterOutlet` properties

BREAKING CHANGE: `RouterOutlet` properties `locationInjector` and `locationFactoryResolver` have been removed as they were deprecated since v4.

mary-poppins commented:
You can preview d12247b at https://pr18781-d12247b.ngbuilds.io/.

ocombe commented:
This PR changed the default behavior of the router. Most of the type the config option `initialNavigation` is `undefined` because there is no default value.

Before the change:
- `undefined`, `legacy_enabled` and `true` had the same behavior: start the initial navigation before the preactivation
- `legacy_disabled` and `false` would just skip the initial navigation
- `enabled` would wait for preactivation, then run the preactivation hooks, and then start the initial navigation
- `disabled` would wait for preactivation and then skip initial navigation

After the change:
- `legacy_enabled`, `legacy_disabled`, `true` and `false` don't work anymore and should be changed to either `enabled` or `disabled`
- `undefined` (which is the default value) now has the same behavior as `enabled`
- `enabled` and `disabled` have the same behaviors as before

To understand the concept of preactivation, read the following design doc: https://docs.google.com/document/d/1Hlw1fPaVs-PCj5KPeJRKhrQGAvFOxdvTlwAcnZosu5A/edit

What does the change imply? Most of the time `initialNavigation` is `undefined`, which means that before the change it didn't run the preactivation before the initial navigation, and it didn't run the preactivation hooks either.
Now it does both of those, and if you get an error with a missing route it will just break and not load the website. That's what happened with one of the test in aio: the test asked protractor to load `index.3.html`, which would trigger an error in the app because the route `index.3.html` didn't exist (the base href was set to `/`, and the default route was empty: `''`). But since it resolved anyway, it just redirected the app to the default route `''` and it worked even with the error. After the change it just stopped at the error, I changed the base href to `/index.3.html` and the test was fixed. This change helped us discover this bug in the application where the error was just ignored, which was wrong. I don't know if there are other things that could break because of this, but I think that the previous behavior was wrong anyway.

ocombe commented:
@jasonaden I updated the commit message, could you check if it's ok?

mary-poppins commented:
You can preview 6e8c01c at https://pr18781-6e8c01c.ngbuilds.io/.

mary-poppins commented:
You can preview 79bbd30 at https://pr18781-79bbd30.ngbuilds.io/.

ocombe commented:
I changed the behavior of this PR to help with the g3 migration: it no longer removed the `legacy_enabled` and `legacy_disabled` options, but the default behavior has still been changed to be `enabled`. Previously we both removed those options and changed the default behavior, forcing teams to update their code immediately.
Now they can set the value to `legacy_enabled` (which was the previous default) and update their code later. It will print a warning in dev mode to prompt them to update.

mary-poppins commented:
You can preview 869df0e at https://pr18781-869df0e.ngbuilds.io/.

mhevery commented:
This is an improvement: http://test/OCL:167316942:BASE:167316961:1504297779950:584f06f0

@jasonaden can you look at this when you get a chance.

matsko commented:
Landed as https://github.com/angular/angular/commit/a9ef8583596a35c422ea51d8a48552830696d836
